renderer/GLSLProgram.cpp

```diff
@@ -0,0 +1,334 @@
+			includedFiles.push_back( fileToInclude );
+			ResolveIncludes( includeContents, includedFiles );
+
+			// also add a #line instruction at beginning and end of include so that
+			// compile errors are mapped to the correct file and line
+			// unfortunately, #line does not take an actual filename, but only an integral reference to a file :(
+			unsigned int currentLine = std::count( source.cbegin(), match[ 0 ].first, '\n' ) + 1 - totalIncludedLines;
+			std::string includeBeginMarker = "#line 0 " + std::to_string( nextFileNo ) + '\n';
+			std::string includeEndMarker = "\n#line " + std::to_string( currentLine ) + ' ' + std::to_string( currentFileNo );
+			totalIncludedLines += std::count( includeContents.begin(), includeContents.end(), '\n' ) + 2;
+
+			// replace include statement with content of included file
+			source.replace( match[ 0 ].first, match[ 0 ].second, includeBeginMarker + includeContents + includeEndMarker );
+		} else {
+			source.replace( match[ 0 ].first, match[ 0 ].second, "// already included " + fileToInclude );
+		}
+	}
+}
+
+/**
+ * Resolves dynamic defines statements in GLSL source files.
+ * Note that the parsing is primitive and not context-sensitive. It will not respect multi-line comments
+ * or conditional preprocessor blocks!
+ * 
+ * Define directives should look like this:
+ * 
+ * #pragma tdm_define "DEF_NAME" // optional comment
+ * 
+ * If DEF_NAME is contained in defines, the line will be replaced by
+ * #define DEF_NAME <value>
+ * 
+ * Otherwise, it will be commented out.
+ */
+void ResolveDefines( std::string &source, const idDict &defines ) {
+	static const std::regex defineRegex( R"regex(^[ \t]*#[ \t]*pragma[ \t]+tdm_define[ \t]+"(.*)"[ \t]*(?:\/\/.*)?\r?$)regex" );
+	
+	std::smatch match;
+	while( std::regex_search( source, match, defineRegex ) ) {
+		std::string define( match[ 1 ].first, match[ 1 ].second );
+		auto defIt = defines.FindKey( define.c_str() );
+		if( defIt != nullptr ) {
+			std::string replacement = "#define " + define + " " + defIt->GetValue().c_str();
+			source.replace( match[ 0 ].first, match[ 0 ].second, replacement );
+		} else {
+			std::string replacement = "// #undef " + define;
+			source.replace( match[ 0 ].first, match[ 0 ].second, replacement );
+		}
+	}
+}
+
+GLuint GLSLProgramLoader::CompileShader( GLint shaderType, const char *sourceFile, const idDict &defines ) {
+	std::string source = ReadFile( sourceFile );
+	if( source.empty() ) {
+		return 0;
+	}
+
+	std::vector<std::string> sourceFiles { sourceFile };
+	ResolveIncludes( source, sourceFiles );
+	ResolveDefines( source, defines );
+
+	GLuint shader = qglCreateShader( shaderType );
+	GLint length = source.size();
+	const char *sourcePtr = source.c_str();
+	qglShaderSource( shader, 1, &sourcePtr, &length );
+	qglCompileShader( shader );
+
+	// check if compilation was successful
+	GLint result;
+	qglGetShaderiv( shader, GL_COMPILE_STATUS, &result );
+	if( result == GL_FALSE ) {
```

Comment:
I wish warnings were displayed even when compilation is successful.

It is not viable right now (because there would be too many warnings), but I hope it will be done somewhere in the bright future =)

Reply:
Do GLSL compilers actually produce useful warnings? We could put output behind a cvar, so it's off by default, but accessible for shader code writing and debugging.